site/Makefile

```diff
@@ -0,0 +1,79 @@
+# Credit to the Kubernetes Website team. (https://github.com/kubernetes/website/blob/main/Makefile)
+HUGO_VERSION      = $(shell grep ^HUGO_VERSION netlify.toml | tail -n 1 | cut -d '=' -f 2 | tr -d " \"\n")
+NODE_BIN          = node_modules/.bin
+NETLIFY_FUNC      = $(NODE_BIN)/netlify-lambda
+
+# The CONTAINER_ENGINE variable is used for specifying the container engine. By default 'docker' is used
+# but this can be overridden when calling make, e.g.
+# CONTAINER_ENGINE=podman make container-image
+CONTAINER_ENGINE ?= docker
+IMAGE_REGISTRY ?= gcr.io/k8s-staging-sig-docs
+IMAGE_VERSION=$(shell scripts/hash-files.sh Dockerfile Makefile | cut -c 1-12)
+CONTAINER_IMAGE   = $(IMAGE_REGISTRY)/kustomize-website-hugo:v$(HUGO_VERSION)-$(IMAGE_VERSION)
+# Mount read-only to allow use with tools like Podman in SELinux mode
+# Container targets don't need to write into /src
+CONTAINER_RUN     = "$(CONTAINER_ENGINE)" run --rm --interactive --tty --volume "$(abspath $(CURDIR)/..):/src:ro,Z"
+
+CCRED=\033[0;31m
+CCEND=\033[0m
+
+# Docker buildx related settings for multi-arch images
+DOCKER_BUILDX ?= docker buildx
+
+.PHONY: help
+help: ## Show this help.
+	@awk 'BEGIN {FS = ":.*?## "} /^[a-zA-Z_-]+:.*?## / {sub("\\\\n",sprintf("\n%22c"," "), $$2);printf "\033[36m%-20s\033[0m %s\n", $$1, $$2}' $(MAKEFILE_LIST)
+
+.PHONY: module-check
+module-check: ## Check if all of the required submodules are correctly initialized.
+	@git submodule status --recursive | awk '/^[+-]/ {err = 1; printf "\033[31mWARNING\033[0m Submodule not initialized: \033[34m%s\033[0m\n",$$2} END { if (err != 0) print "You need to run \033[32mmake module-init\033[0m to initialize missing modules first"; exit err }' 1>&2
+
+.PHONY: module-init
+module-init: ## Initialize required submodules.
+	@echo "Initializing submodules..." 1>&2
+	@git submodule update --init --recursive --depth 1
+
+.PHONY: all
+all: build ## Build site with production settings and put deliverables in ./public
+
+.PHONY: build
+build: module-check ## Build site with non-production settings and put deliverables in ./public
+	hugo --cleanDestinationDir --minify --environment development
+
+.PHONY: build-preview
+build-preview: module-check ## Build site with drafts and future posts enabled
+	hugo --cleanDestinationDir --buildDrafts --buildFuture --environment preview
+
+.PHONY: serve
+serve: module-check ## Boot the development server.
+	hugo server --buildFuture --environment development
+
+container-image.sentinel: netlify.toml Dockerfile hugo.toml Makefile
+	$(CONTAINER_ENGINE) build . \
+		--network=host \
+		--tag $(CONTAINER_IMAGE) \
+		--build-arg HUGO_VERSION=$(HUGO_VERSION)
+	@touch $@
```

Comment:
I'm a little curious about this target! Do we need to write a file to disk when building? Could you elaborate on the use case for this (or maybe on the reason for which having this logic in the `container-image` target is undesirable/not ideal)?

Reply:
The `container-image.sentinel` rule only builds the container image if a container image dependency, i.e. a Dockerfile, has been touched after an image build. The sentinel file was added as a dependency to the `container-serve` rule so that a user does not need to remember to build the image when running the site with Docker. The image will only build if an image dependency has changed since the last time an image was built which is tracked by the sentinel file.

A phony target will always execute the rule. The image would always build if I add the image build rule to the `container-image` target and reference this as a dependency.

The sentinel file is meant to be a convenience. Here is a [blog about Makefiles](https://tech.davis-hansson.com/p/make/) that I like to reference. I can remove this logic if you don't like the approach.

Reply:
I'm familiar with `make` 🙂

What I was trying to get at is more around what problem you're trying to solve. I personally don't think rebuilding an image each time you run the target (`container-image`, in this instance) would be an issue, because `container-image` doesn't need to be run each time one issues `container-serve`, and `container-serve` does not depend on `container-image`.

I think your approach has merit, but I can easily break the flow by removing the built image, which is an operation I routinely do to release disk space when I'm done using a given image. But now, there's the aggravating factor that `container-image` will _not_ rebuild the image for me unless I remove the sentinel file -- and the response I get from `make` is not super helpful (`Nothing to be done for 'container-image'.`).

If you feel strongly about keeping it as-is, I'd advise to:
* Include a bit of documentation explaining what it does for people less familiar with `make`; and
* Include a target to clean up the sentinel file.

Reply:
No worries. I removed the sentinel file. The `container-image` rule will build the image.

Reply:
Feel free to resolve this thread too 🙂